File: app/components/folio/console/files/display_toggle_component.rb
# frozen_string_literal: true

class Folio::Console::Files::DisplayToggleComponent < Folio::Console::ApplicationComponent
  def initialize; end

  def buttons_data
    [
      {
        icon: :grid_view,
        class: "f-c-files-display-toggle__btn f-c-files-display-toggle__btn--enabled-false",
        data: stimulus_action({ click: "click" }, { enabled: false })
      },
      {
        icon: :format_list_bulleted,
        class: "f-c-files-display-toggle__btn f-c-files-display-toggle__btn--enabled-true",
        data: stimulus_action({ click: "click" }, { enabled: true })
      }
    ]
  end

  def data
    enabled = if Folio::Current.user && Folio::Current.user.console_preferences.present?
      Folio::Current.user.console_preferences["images_table_view"].present?
    else
      false
    end

    stimulus_controller("f-c-files-display-toggle",
                        values: {
                          api_url: controller.update_console_preferences_console_api_current_user_path,
                          enabled:
                        })
  end
end

File: app/components/folio/console/files/display_toggle_component.js
window.Folio.Stimulus.register('f-c-files-display-toggle', class extends window.Stimulus.Controller {
  static values = {
    apiUrl: String,
    enabled: Boolean
  }

  click (e) {
    this.enabledValue = e.params.enabled

    for (const toggle of document.querySelectorAll('.f-c-files-display-toggle')) {
      if (toggle !== this.element) {
        toggle.dataset.fCFilesDisplayToggleEnabledValue = String(this.enabledValue)
      }
    }

    for (const list of document.querySelectorAll('.f-file-list--view-changeable')) {
      list.dispatchEvent(new CustomEvent('f-c-files-display-toggle:table-view-change', { bubbles: true, detail: { images_table_view: this.enabledValue } }))
    }

    const data = { images_table_view: this.enabledValue }
    window.Folio.Api.apiPost(this.apiUrlValue, data)
  }
})
